src/outputs/restart_opmd.cpp

```diff
@@ -0,0 +1,180 @@
+  } else {
+    return -1;
+  }
+}
+
+RestartReaderOPMD::SparseInfo RestartReaderOPMD::GetSparseInfo() const {
+  // TODO(pgrete) needs impl
+  return {};
+}
+
+RestartReaderOPMD::MeshInfo RestartReaderOPMD::GetMeshInfo() const {
+  RestartReaderOPMD::MeshInfo mesh_info;
+  mesh_info.nbnew = it->getAttribute("NBNew").get<int>();
+  mesh_info.nbdel = it->getAttribute("NBDel").get<int>();
+  mesh_info.nbtotal = it->getAttribute("NumMeshBlocks").get<int>();
+  mesh_info.root_level = it->getAttribute("RootLevel").get<int>();
+
+  mesh_info.bound_cond =
+      it->getAttribute("BoundaryConditions").get<std::vector<std::string>>();
+
+  mesh_info.block_size = it->getAttribute("MeshBlockSize").get<std::vector<int>>();
+  mesh_info.includes_ghost = it->getAttribute("IncludesGhost").get<int>();
+  mesh_info.n_ghost = it->getAttribute("NGhost").get<int>();
+
+  mesh_info.grid_dim = it->getAttribute("RootGridDomain").get<std::vector<Real>>();
+  mesh_info.lx123 = it->getAttribute("loc.lx123").get<std::vector<int64_t>>();
+  mesh_info.level_gid_lid_cnghost_gflag =
+      it->getAttribute("loc.level-gid-lid-cnghost-gflag").get<std::vector<int>>();
+
+  return mesh_info;
+}
+
+SimTime RestartReaderOPMD::GetTimeInfo() const {
+  SimTime time_info{};
+
+  time_info.time = it->time<Real>();
+  time_info.dt = it->dt<Real>();
+  time_info.ncycle = it->getAttribute("NCycle").get<int>();
+
+  return time_info;
+}
+// Gets the counts and offsets for MPI ranks for the meshblocks set
+// by the indexrange. Returns the total count on this rank.
+std::size_t RestartReaderOPMD::GetSwarmCounts(const std::string &swarm,
+                                              const IndexRange &range,
+                                              std::vector<std::size_t> &counts,
+                                              std::vector<std::size_t> &offsets) {
+  // TODO(pgrete) needs impl
+  return 0;
+}
+
+template <typename T>
+void RestartReaderOPMD::ReadAllParamsOfType(const std::string &pkg_name, Params &params) {
+  for (const auto &key : params.GetKeys()) {
+    const auto type = params.GetType(key);
+    auto mutability = params.GetMutability(key);
+    if (type == std::type_index(typeid(T)) && mutability == Params::Mutability::Restart) {
+      auto val = it->getAttribute("Params/" + pkg_name + "/" + key).get<T>();
+      params.Update(key, val);
+    }
+  }
+}
+
+template <typename... Ts>
+void RestartReaderOPMD::ReadAllParamsOfMultipleTypes(const std::string &pkg_name,
+                                                     Params &p) {
+  ([&] { ReadAllParamsOfType<Ts>(pkg_name, p); }(), ...);
+}
+
+template <typename T>
+void RestartReaderOPMD::ReadAllParams(const std::string &pkg_name, Params &p) {
+  ReadAllParamsOfMultipleTypes<T, std::vector<T>>(pkg_name, p);
+  // TODO(pgrete) check why this doens't work, i.e., which type is causing problems
+  // ReadAllParamsOfMultipleTypes<PARTHENON_ATTR_VALID_VEC_TYPES(T)>(pkg, it);
```

Comment:
It was subtle to get this to work for hdf5. Maybe some template type resolution failure?